sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+        s"The given input maps of function $prettyName should all be the same type, " +
+          "but they are " + children.map(_.dataType.simpleString).mkString("[", ", ", "]"))
+    } else if (children.map(_.dataType.asInstanceOf[MapType].valueType).distinct.length > 1) {
+      TypeCheckResult.TypeCheckFailure(
+        s"The given input maps of function $prettyName should all be the same type, " +
+          "but they are " + children.map(_.dataType.simpleString).mkString("[", ", ", "]"))
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def dataType: MapType = {
+    MapType(
+      keyType = children.headOption
+        .map(_.dataType.asInstanceOf[MapType].keyType).getOrElse(StringType),
+      valueType = children.headOption
+        .map(_.dataType.asInstanceOf[MapType].valueType).getOrElse(StringType),
+      valueContainsNull = children.map { c =>
+        c.dataType.asInstanceOf[MapType]
+      }.exists(_.valueContainsNull)
+    )
+  }
+
+  override def nullable: Boolean = children.exists(_.nullable)
+
+  override def eval(input: InternalRow): Any = {
+    val union = new util.LinkedHashMap[Any, Any]()
+    children.map(_.eval(input)).foreach { raw =>
+      if (raw == null) {
+        return null
+      }
+      val map = raw.asInstanceOf[MapData]
+      map.foreach(dataType.keyType, dataType.valueType, (k, v) =>
+        union.put(k, v)
+      )
+    }
+    val (keyArray, valueArray) = union.entrySet().toArray().map { e =>
+      val e2 = e.asInstanceOf[java.util.Map.Entry[Any, Any]]
+      (e2.getKey, e2.getValue)
+    }.unzip
+    new ArrayBasedMapData(new GenericArrayData(keyArray), new GenericArrayData(valueArray))
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val mapCodes = children.map(c => c.genCode(ctx))
+    val keyType = dataType.keyType
+    val valueType = dataType.valueType
+    val mapRefArrayName = ctx.freshName("mapRefArray")
+    val unionMapName = ctx.freshName("union")
+
+    val mapDataClass = classOf[MapData].getName
+    val arrayBasedMapDataClass = classOf[ArrayBasedMapData].getName
+    val arrayDataClass = classOf[ArrayData].getName
+    val genericArrayDataClass = classOf[GenericArrayData].getName
+    val hashMapClass = classOf[util.LinkedHashMap[Any, Any]].getName
+    val entryClass = classOf[util.Map.Entry[Any, Any]].getName
+
+    val init =
+      s"""
+        |$mapDataClass[] $mapRefArrayName = new $mapDataClass[${mapCodes.size}];
+        |boolean ${ev.isNull} = false;
+        |$mapDataClass ${ev.value} = null;
+      """.stripMargin
+
+    val assignments = mapCodes.zipWithIndex.map { case (m, i) =>
+      val initCode = mapCodes(i).code
+      val valueVarName = mapCodes(i).value.code
+      s"""
+         |$initCode
+         |$mapRefArrayName[$i] = $valueVarName;
+         |if ($valueVarName == null) {
+         |  ${ev.isNull} = true;
+         |}
+       """.stripMargin
+    }.mkString("\n")
```

Comment:
Use `ctx.splitExpressionsWithCurrentInputs()` or something to avoid exceeding JVM limit.